staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/customresource_handler.go

```diff
@@ -919,6 +928,51 @@ func (r *crdHandler) getOrCreateServingInfoFor(uid types.UID, name string) (*crd
 			}
 		}
 	}
+	if utilfeature.DefaultFeatureGate.Enabled(features.StorageVersionAPI) &&
+		utilfeature.DefaultFeatureGate.Enabled(features.APIServerIdentity) {
+		// Update storage version records before returning the new handler.
+		// This means stale records can still exist in the cluster if an apiserver
+		// hasn't got any request. All apiservers have to at least serve one
+		// request (can be a GET) for a CR's storage version record to be
+		// up-to-date, and to trigger storage migration.
```

Comment:
An alternative to the owner-reference approach is to let `crdHandler` garbage collect storageversion whenever we tear down old storage. That approach is more responsive, but it's more expensive (apiextensions-apiserver replicas conflict with each other in HA cluster) and only helps with some rare corner cases.

Reply:
Specifically, in an HA cluster, after a CRD gets encoding version update:

1. if all three server have served requests with new storage, storage migration will be triggered for both approaches. 
2. if only one or two of the three servers have served requests with new storage, storage migration won't happen for the owner-reference approach, but will for the alternative
3. if none of the three servers has served requests with new storage, storage migration won't happen for the owner-reference approach; it will for the alternative **by a chance**: 
  a. if the storage migrator observed the old StorageVersion object getting shrunk to one/two server records, it will trigger the migration
  b. if the watch events get compressed (e.g. informer reconnection) and the storage migrator only saw a DELETE event, it won't trigger the migration

3 (updating the schema without even read after the update) should be extremely rare. I don't expect 2 to happen assuming enough number (>3) of requests have made to the servers and the load balancer works reasonably. 

That said, we can let the storage migrator send read requests periodically to bump the chances of 1 & 2. Also the alternative can be a future improvement if one region being starved is a concern.

Reply:
Having this logic in the handler chain feels misplaced. Why is that? Just register in the event handlers for CRDs (where we also install the handler in the first place).

Reply:
Also above we have event handlers: https://github.com/kubernetes/kubernetes/pull/96403/files#diff-01e7a2059fe62e95ec32e2589c01b7218100de327647c641c334d38f2c5aa06eL206.

Reply:
not sure what is `pkg/controllers/storageversion`. 

let me think about using the event handlers to create/update storage versions. I don't want to use the event handlers to garbage collect storage versions because of reasons in https://github.com/kubernetes/kubernetes/pull/96403/files#r520386107.

Reply:
I agree using the event handlers to create/update storage versions works better, and solves https://github.com/kubernetes/kubernetes/pull/96403/files#r520403155. Will update the PR